### PR TITLE
Fix the Suggestions Database Updates Handling

### LINKED ISSUE #1141: Completion suggestions contains duplicated values or non-available locals
### General Summary

During initial testing of searcher I got some duplicated entries:
image
As I checked, the view reflect exactly the engine's response.

### Steps to Reproduce

Open the attached project in project manager:
[project.zip](https://github.com/enso-org/enso/files/5207421/project.zip)

Then ask for suggestion database, and for completion:
```
{"jsonrpc":"2.0","id":1,"method":"search/getSuggestionsDatabase","params":{}}
{"jsonrpc":"2.0","id":9,"method":"search/completion","params":{"file":{"rootId":"b481adbf-6de2-4267-8aa6-7f4238c6d84b","segments":["src","Main.enso"]},"position":{"character":7,"line":6},"returnType":null,"selfType":null,"tags":null}}
```
(replace the rootId of course).

### Expected Result

Got completion response with Main.foo, Main.main, a, b, c

### Actual Result

Got response with three a, two b, and two c.

the result of getSuggestionsDatabase:
```
{"jsonrpc":"2.0","id":1,"result":{"entries":[{"id":890,"suggestion":{"type":"local","externalId":"70e4d3d8-c292-438b-96b5-10a8c723263c","module":"Unnamed.Main","name":"a","returnType":"Number","scope":{"start":{"line":2,"character":7},"end":{"line":6,"character":22}}}},{"id":891,"suggestion":{"type":"local","externalId":"c256a70a-2b4c-43ff-98c4-fcaba739a784","module":"Unnamed.Main","name":"b","returnType":"Number","scope":{"start":{"line":2,"character":7},"end":{"line":6,"character":22}}}},{"id":853,"suggestion":{"type":"method","externalId":"c4a6e583-dd90-4f32-b5a7-11bd4f2a2d6d","module":"Unnamed.Main","name":"foo","arguments":[{"name":"this","reprType":"Any","isSuspended":false,"hasDefault":false,"defaultValue":null},{"name":"a","reprType":"Any","isSuspended":false,"hasDefault":false,"defaultValue":null},{"name":"b","reprType":"Any","isSuspended":false,"hasDefault":false,"defaultValue":null}],"selfType":"Main","returnType":"Any"}},{"id":908,"suggestion":{"type":"local","externalId":"d4dd9087-7b31-44c5-a01b-766672a7208b","module":"Unnamed.Main","name":"c","returnType":"Number","scope":{"start":{"line":2,"character":7},"end":{"line":6,"character":11}}}},{"id":907,"suggestion":{"type":"local","externalId":"c256a70a-2b4c-43ff-98c4-fcaba739a784","module":"Unnamed.Main","name":"b","returnType":"Number","scope":{"start":{"line":2,"character":7},"end":{"line":6,"character":11}}}},{"id":924,"suggestion":{"type":"local","externalId":"d4dd9087-7b31-44c5-a01b-766672a7208b","module":"Unnamed.Main","name":"c","returnType":"Number","scope":{"start":{"line":2,"character":7},"end":{"line":6,"character":8}}}},{"id":906,"suggestion":{"type":"local","externalId":"70e4d3d8-c292-438b-96b5-10a8c723263c","module":"Unnamed.Main","name":"a","returnType":"Number","scope":{"start":{"line":2,"character":7},"end":{"line":6,"character":11}}}},{"id":923,"suggestion":{"type":"local","externalId":"c256a70a-2b4c-43ff-98c4-fcaba739a784","module":"Unnamed.Main","name":"b","returnType":"Number","scope":{"start":{"line":2,"character":7},"end":{"line":6,"character":8}}}},{"id":892,"suggestion":{"type":"local","externalId":"d4dd9087-7b31-44c5-a01b-766672a7208b","module":"Unnamed.Main","name":"c","returnType":"Number","scope":{"start":{"line":2,"character":7},"end":{"line":6,"character":22}}}},{"id":925,"suggestion":{"type":"method","externalId":"83ffcb0b-6edd-4694-8b34-fbd92114366b","module":"Unnamed.Main","name":"main","arguments":[{"name":"this","reprType":"Any","isSuspended":false,"hasDefault":false,"defaultValue":null}],"selfType":"Main","returnType":"Any"}},{"id":922,"suggestion":{"type":"local","externalId":"70e4d3d8-c292-438b-96b5-10a8c723263c","module":"Unnamed.Main","name":"a","returnType":"Number","scope":{"start":{"line":2,"character":7},"end":{"line":6,"character":8}}}}],"currentVersion":1916}}
```
The result of getting completion suggestions:
```
{"jsonrpc":"2.0","id":9,"result":{"results":[853,890,891,892,906,907,908,922,923,924,925],"currentVersion":1917}}
```

### Enso Version

7373da2 (current main)

### PR DESCRIPTION
**Pull request by 4e6**
### Pull Request Description

close #1141 

Changelog:

- fix: update the cached file version when applying the suggestion updates
- refactor: simplify the runtime protocol about notifications in the suggestion updates

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/main/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/main/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guides.
- [x] All documentation and configuration conforms to the [markdown](https://github.com/enso-org/enso/blob/main/docs/style-guide/markdown.md) and [YAML](https://github.com/enso-org/enso/blob/main/docs/style-guide/yaml.md) style guides.
- [x] All code has been tested where possible.

### CODE REVIEW

File: engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala [hunk 1]
```diff
@@ -386,28 +299,40 @@ final class SuggestionsHandler(
...
-    val (added, removed) = msg.updates
-      .foldLeft((Seq[Suggestion](), Seq[Suggestion]())) {
-        case ((add, remove), msg: Api.SuggestionsDatabaseUpdate.Add) =>
-          (add :+ msg.suggestion, remove)
-        case ((add, remove), msg: Api.SuggestionsDatabaseUpdate.Remove) =>
-          (add, remove :+ msg.suggestion)
+    val (addCmds, removeCmds, cleanCmds) = msg.updates
+      .foldLeft((Seq[Suggestion](), Seq[Suggestion](), Seq[String]())) {
+        case ((add, remove, clean), m: Api.SuggestionsDatabaseUpdate.Add) =>
+          (add :+ m.suggestion, remove, clean)
```
Comment (kustosz) [inline engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala:309]: This is going to be slow for larger modules, use `::` and reverse at the end.

Comment (kustosz) [reply engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala:309]: Same for all other branches.

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/job/EnsureCompiledJob.scala [hunk 2]
```diff
@@ -92,33 +92,40 @@ class EnsureCompiledJob(protected val files: Iterable[File])
...
   }
 
   private def analyzeImport(
     module: Module
   )(implicit ctx: RuntimeContext): Unit = {
-    if (!module.isIndexed && module.getLiteralSource != null) {
+    if (
+      !module.isIndexed &&
+      module.getLiteralSource != null &&
+      module.getPath != null
```
Comment (kustosz) [inline engine/runtime/src/main/scala/org/enso/interpreter/instrument/job/EnsureCompiledJob.scala:107]: Not sure, but I think it is possible for the IDE to create an unsaved module? So we shouldn't really depend on the module being on-disk.

File: engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala [hunk 3]
```diff
@@ -946,9 +956,11 @@ class RuntimeServerTest
       context.Main.Update.mainZ(contextId),
       idMainUpdate,
       Api.Response(
-        Api.SuggestionsDatabaseReIndexNotification(
-          moduleName,
+        Api.SuggestionsDatabaseModuleUpdateNotification(
```
Comment (kustosz) [inline engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala:959]: Why does a suggestion DB carry the whole file in it? This seems very wasteful

**Review by kustosz — APPROVED** [review]
I've raised a few issues inline, but I don't know enough about this code to be sure they are valid. So approval conditional on them being either invalid or fixed :)

File: engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala (hunk 1 above)
Comment (4e6) [reply engine/language-server/src/main/scala/org/enso/languageserver/search/SuggestionsHandler.scala:309]: Then, it's better to use vectors here. For some reason, I decided that the default implementation of `Seq` is a `Vector`.

File: engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala (hunk 3 above)
Comment (4e6) [reply engine/runtime/src/test/scala/org/enso/interpreter/test/instrument/RuntimeServerTest.scala:959]: I also planned to address this in #1162, as the fix will require a bit of refactoring :+1:

File: engine/runtime/src/main/scala/org/enso/interpreter/instrument/job/EnsureCompiledJob.scala (hunk 2 above)
Comment (4e6) [reply engine/runtime/src/main/scala/org/enso/interpreter/instrument/job/EnsureCompiledJob.scala:107]: Good catch. I added this case to the #1162 specification, which I'll do next.

**Review by lolczak — APPROVED** [review]

**Pull request merged**